internal/concurrent/future.go

```diff
@@ -0,0 +1,76 @@
+package concurrent
+
+import "context"
+
+// Future are a way to easily start a go routine and get the result of the go routine later. Avoid building any
+// mapping functionality into futures, as that can make them difficult to reason about
+type (
+	result[T any] struct {
+		err error
+		res T
+	}
+
+	GoRoutineRunner interface {
+		// Go starts a go routine and returns an error if the go routine could not be started.
+		Go(context.Context, func()) error
```

Comment:
Is there ever a scenario where the goroutine cannot be started? Is it for an already canceled/completed context? (in which case SynchronousGoRoutineRunner should probably behave the same?)

Reply:
Ah I see GoRoutineLimter now... disregard ^

Reply:
Well hmm GoRoutineLimter errors on Go on canceled context, so maybe the sync & async runners should too... Let me know what you think

Reply:
Good point! I'll add it!